File: src/main/java/me/glorantq/aboe/chat/client/channels/ClientChannelManager.java
package me.glorantq.aboe.chat.client.channels;

import cpw.mods.fml.common.network.simpleimpl.IMessage;
import cpw.mods.fml.common.network.simpleimpl.IMessageHandler;
import cpw.mods.fml.common.network.simpleimpl.MessageContext;
import lombok.Getter;
import me.glorantq.aboe.chat.ABOEChat;
import me.glorantq.aboe.chat.common.PacketChangeChannel;
import me.glorantq.aboe.chat.common.PacketChangeChannelResponse;
import me.glorantq.aboe.chat.common.PacketChannelChanged;
import me.glorantq.aboe.chat.common.PacketChannelList;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.List;

public class ClientChannelManager {
    private final Logger logger = LogManager.getLogger("ClientChannelManager");

    private final ABOEChat aboeChat = ABOEChat.getInstance();

    private @Getter String currentChannel = "#unknown";
    private final  @Getter List<ClientChatChannel> availableChannels = new ArrayList<>();

    private void onChannelChanged(String newChannel) {
        this.currentChannel = newChannel;
        logger.info("Channel changed to: {}", newChannel);
    }

    private void onChannelStatusChanged(List<ClientChatChannel> changed) {
        synchronized (availableChannels) {
            availableChannels.clear();
            availableChannels.addAll(changed);
        }
    }

    private void onChangeChannelResponse(boolean successful, String extraMessage, List<ClientChatChannel> changedChannels) {
        System.out.println("Change: " + successful + "  " + extraMessage + "  " + changedChannels.size());
    }

    public void changeToChannel(String channelName) {
        aboeChat.getSimpleNetworkWrapper().sendToServer(new PacketChangeChannel(channelName));
    }

    public static class PacketChannelChangedHandler implements IMessageHandler<PacketChannelChanged, IMessage> {
        private final ClientChannelManager clientChannelManager = ABOEChat.getInstance().getClientChannelManager();

        @Override
        public IMessage onMessage(PacketChannelChanged message, MessageContext ctx) {
            clientChannelManager.onChannelChanged(message.getNewChannel());

            return null;
        }
    }

    public static class PacketChannelListHandler implements IMessageHandler<PacketChannelList, IMessage> {
        private final ClientChannelManager clientChannelManager = ABOEChat.getInstance().getClientChannelManager();

        @Override
        public IMessage onMessage(PacketChannelList message, MessageContext ctx) {
            clientChannelManager.onChannelStatusChanged(message.getChannels());

            return null;
        }
    }

    public static class PacketChangeChannelResponseHandler implements IMessageHandler<PacketChangeChannelResponse, IMessage> {
        private final ClientChannelManager clientChannelManager = ABOEChat.getInstance().getClientChannelManager();

        @Override
        public IMessage onMessage(PacketChangeChannelResponse message, MessageContext ctx) {
            clientChannelManager.onChangeChannelResponse(message.isSuccessful(), message.getExtraMessage(), message.getChangedChannels());

            return null;
        }
    }
}

File: src/main/kotlin/me/glorantq/aboe/chat/client/channels/ClientChannelManager.kt
package me.glorantq.aboe.chat.client.channels

import cpw.mods.fml.common.network.simpleimpl.IMessage
import cpw.mods.fml.common.network.simpleimpl.IMessageHandler
import cpw.mods.fml.common.network.simpleimpl.MessageContext
import me.glorantq.aboe.chat.ABOEChat
import me.glorantq.aboe.chat.ClientChannelList
import me.glorantq.aboe.chat.common.PacketChangeChannel
import me.glorantq.aboe.chat.common.PacketChangeChannelResponse
import me.glorantq.aboe.chat.common.PacketChannelChanged
import me.glorantq.aboe.chat.common.PacketChannelList
import org.apache.logging.log4j.LogManager
import java.util.*

class ClientChannelManager {
    private val logger = LogManager.getLogger("ClientChannelManager")

    private val aboeChat: ABOEChat = ABOEChat.instance
    var currentChannel = "#unknown"
        private set
    val availableChannels: ClientChannelList = ArrayList()

    private fun onChannelChanged(newChannel: String) {
        this.currentChannel = newChannel
        logger.info("Channel changed to: {}", newChannel)
    }

    private fun onChannelStatusChanged(changed: List<ClientChatChannel>) {
        synchronized(availableChannels) {
            availableChannels.clear()
            availableChannels.addAll(changed)
        }
    }

    private fun onChangeChannelResponse(successful: Boolean, extraMessage: String, changedChannels: List<ClientChatChannel>) {
        println("Change: " + successful + "  " + extraMessage + "  " + changedChannels.size)
    }

    fun changeToChannel(channelName: String) {
        aboeChat.simpleNetworkWrapper!!.sendToServer(PacketChangeChannel(channelName))
    }

    class PacketChannelChangedHandler : IMessageHandler<PacketChannelChanged, IMessage> {
        private val clientChannelManager: ClientChannelManager? = ABOEChat.instance.clientChannelManager

        override fun onMessage(message: PacketChannelChanged, ctx: MessageContext): IMessage? {
            clientChannelManager!!.onChannelChanged(message.newChannel)

            return null
        }
    }

    class PacketChannelListHandler : IMessageHandler<PacketChannelList, IMessage> {
        private val clientChannelManager: ClientChannelManager? = ABOEChat.instance.clientChannelManager

        override fun onMessage(message: PacketChannelList, ctx: MessageContext): IMessage? {
            clientChannelManager!!.onChannelStatusChanged(message.channels)

            return null
        }
    }

    class PacketChangeChannelResponseHandler : IMessageHandler<PacketChangeChannelResponse, IMessage> {
        private val clientChannelManager: ClientChannelManager? = ABOEChat.instance.clientChannelManager

        override fun onMessage(message: PacketChangeChannelResponse, ctx: MessageContext): IMessage? {
            clientChannelManager!!.onChangeChannelResponse(message.successful, message.extraMessage, message.changedChannels)

            return null
        }
    }
}
